### DISCUSSION TITLE
Plotting ground state populations of instantaneous Hamiltonian

### DISCUSSION THREAD

**Discussion by qflux98** [qucontrol/krotov#104] [Q&A] [answered]
Instead of plotting state populations w.r.t some fixed operator as described in several Krotov example notebooks (e.g. https://qucontrol.github.io/krotov/v1.0.0/notebooks/01_example_simple_state_to_state.html), how can I plot ground state populations for instantaneous hamiltonian for every time grids after pulse controls are optimized? Do you have relevant example notebooks?

**Comment by goerz** [comment, marked as answer]
You'd have to sum the nested list that encodes the time-dependent Hamiltonian into a single `Qobj`, and then call QuTiP's `eigenstates` method to find the ground state, at every time step.

We don't have an example that does this.

**Reply by qflux98** [reply]
so we should pull out our optimized pulse list, make a list that encodes Hamiltonian w.r.t the pulse list, and find ground states in each time grid. Is there any more convenient way, i.e. using `mesolve`?

**Reply by goerz** [reply]
That's right. You'll have to get the ground state `|Ψ₀(t)⟩` at each point in time first. Once you have those, you can write a function `exp_ground(Ψ, t)` that calculates the expectation value of `|Ψ₀(t)⟩⟨Ψ₀(t)|` for a given `|Ψ(t)⟩`. You can then pass that to `mesolve`, using that an element of `e_ops` can be a callback function [according to the documentation](https://qutip.org/docs/latest/apidoc/functions.html?highlight=callback#qutip.mesolve.mesolve)

**Reply by goerz** [reply]
I can't think of any convenience function that would help you with getting the `|Ψ₀(t)⟩`, but calculating those should be pretty straightforward, albeit slow for larger Hilbert space dimensions

**Reply by qflux98** [reply]
okay, is `opt_result.objectives[0].mesolve` supposed to simulate the hamiltonian with optimized pulse sequence?

**Reply by qflux98** [reply]
as an additional question, is there a way to always start from an initial state to be a ground state of H(0) while continuing the optimization procedure from the previous optimization by using `continue_from=opt_result`? I tried to change the initial state after one optimization is done and used `continue_from=opt_result` to continue, but then an error occurred because it changed the objective.

**Reply by goerz** [reply]
If instead of `opt_result.objectives[0].mesolve`, you meant to write `opt_result.optimized_objectives[0].mesolve`, then yes, that would run `qutip.mesolve` using the initial state and the Hamiltonian from the objective (with the optimized fields in the second case).

I'm not sure what you mean in the second question. Isn't `H(0)` the same in both optimizations?

In general, if the objectives change, you can't use `continue_from`. That option is for when you want to continue a previous optimization with modified convergence criteria / `iter_max`.

**Reply by qflux98** [reply]
are we supposed to fix the boundary conditions of pulse sequence throughout the optimization (in the conventional field of QOC)? If not, I was thinking of a case where your pulse value at t=0 may vary after the first optimization by using some S(t) that is non-zero at t=0 (e.g. S(t) = 1).

**Reply by goerz** [reply]
Changing the initial state in each iteration like that is not supported. You might be able to get it to work somehow, either via hooks or by doing only a single iteration for each call of optimize_pulses. It seems like a very questionable approach to me, but maybe I’m wrong. In any case, your life will be much simpler if you keep S(t) = 0 at the boundaries in this scenario

**Reply by qflux98** [reply]
You're absolutely right. In fact, my question is coming from wondering whether for a very large annealing time T, if you don't change your final value of pulse schedule, which means always fixing the final Hamiltonian, then this might cause a training problem due to the adiabaticity of the system (The final state should be in the ground state of the final Hamiltonian that is fixed)
